ext/opentelemetry-ext-asgi/src/opentelemetry/ext/asgi/__init__.py

```diff
@@ -0,0 +1,206 @@
+    if code <= 399:
+        if allow_redirect:
+            return StatusCanonicalCode.OK
+        return StatusCanonicalCode.DEADLINE_EXCEEDED
+    if code <= 499:
+        if code == 401:  # HTTPStatus.UNAUTHORIZED:
+            return StatusCanonicalCode.UNAUTHENTICATED
+        if code == 403:  # HTTPStatus.FORBIDDEN:
+            return StatusCanonicalCode.PERMISSION_DENIED
+        if code == 404:  # HTTPStatus.NOT_FOUND:
+            return StatusCanonicalCode.NOT_FOUND
+        if code == 429:  # HTTPStatus.TOO_MANY_REQUESTS:
+            return StatusCanonicalCode.RESOURCE_EXHAUSTED
+        return StatusCanonicalCode.INVALID_ARGUMENT
+    if code <= 599:
+        if code == 501:  # HTTPStatus.NOT_IMPLEMENTED:
+            return StatusCanonicalCode.UNIMPLEMENTED
+        if code == 503:  # HTTPStatus.SERVICE_UNAVAILABLE:
+            return StatusCanonicalCode.UNAVAILABLE
+        if code == 504:  # HTTPStatus.GATEWAY_TIMEOUT:
+            return StatusCanonicalCode.DEADLINE_EXCEEDED
+        return StatusCanonicalCode.INTERNAL
+    return StatusCanonicalCode.UNKNOWN
+
+
+def collect_request_attributes(scope):
+    """Collects HTTP request attributes from the ASGI scope and returns a
+    dictionary to be used as span creation attributes."""
+
+    port = scope.get("server")[1]
+    server_host = scope.get("server")[0] + (
+        ":" + str(port) if port != 80 else ""
+    )
+    http_url = scope.get("scheme") + "://" + server_host + scope.get("path")
+    if scope.get("query_string"):
+        http_url = http_url + ("?" + scope.get("query_string").decode("utf8"))
+
+    result = {
+        "component": scope.get("type"),
+        "http.method": scope.get("method"),
+        "http.server_name": scope.get("server")[0],
+        "http.scheme": scope.get("scheme"),
+        "http.host": server_host,
+        "host.port": port,
+        "http.flavor": scope.get("http_version"),
+        "http.target": scope.get("path"),
+        "http.url": http_url,
+    }
+
+    if "client" in scope:
+        result["net.peer.ip"] = scope.get("client")[0]
+        result["net.peer.port"] = scope.get("client")[1]
+
+    return result
+
+
+def set_status_code(span, status_code):
+    """Adds HTTP response attributes to span using the status_code argument."""
+    try:
+        status_code = int(status_code)
+    except ValueError:
+        span.set_status(
+            Status(
+                StatusCanonicalCode.UNKNOWN,
+                "Non-integer HTTP status: " + repr(status_code),
+            )
+        )
+    else:
+        span.set_attribute("http.status_code", status_code)
+        span.set_status(Status(http_status_to_canonical_code(status_code)))
+
+
+def get_default_span_name(scope):
+    """Calculates a (generic) span name for an incoming HTTP request based on the ASGI scope."""
+
+    # TODO: Update once
+    #  https://github.com/open-telemetry/opentelemetry-specification/issues/270
+    #  is resolved
+    return scope.get("path", "/")
```

Comment:
looking at the conversation in the opentelemetry-specification, we should use the HTTP method rather than the specialized path, since it has very high cardinality: https://github.com/open-telemetry/opentelemetry-specification/pull/416/files.

But this is a bug with the existing wsgi implementation as well.

Reply:
Here's a tracking ticket for the wsgi implementation, but it'd be good to fix it in ASGI now: https://github.com/open-telemetry/opentelemetry-python/issues/409

Reply:
Acknowledged, I'll change it to utilize the method_name.

Do you feel like I should make a similar change in the WSGI implementation to keep them equivalent, or would you rather have that change seperately?

Reply:
It appears the recent update to the specification includes
> Instrumentation MUST NOT default to using URI
path as span name, but MAY provide hooks to allow custom logic to override the
default span name.
Might I suggest adding an optional callback to this middleware to allow clients to specify a override the default span name?

Reply:
I'll add an optional callback to this, and pass it through.
I will make a similar change to the WSGI implementation in another PR, just to keep them syncronized.

Reply:
Resolved in: https://github.com/open-telemetry/opentelemetry-python/pull/402/commits/ede28b2cc42551cb57c015f3af0d153efbdf814d

Reply:
WSGI equivalent: https://github.com/open-telemetry/opentelemetry-python/pull/440